### PR TITLE
initial port of logic to dotnet standard 1.4

### PR DESCRIPTION
**Pull request by Jetski5822**
This is an initial port to dotnet standard 1.4, so should support your internal platforms interns of backward compatibility... however I have change a 'few' things

1. Added the IOptions api to deal with configuration
2. moved everything to use JsonProperty removing dotnet serialisation properties.
3. Renamed the ZendeskApi.Contracts project to ZendeskApi.Client.Abstractions
4. Removed a lot of nesting on API calls... I moved to a model like this...

```c#
using (_loggerScope(_logger, "GetAllAsync"))
using (var client = _apiClient.CreateClient())
{
    var response = await client.GetAsync(GroupsResourceUri).ConfigureAwait(false);

    response.EnsureSuccessStatusCode();

    return (await response.Content.ReadAsAsync<GroupsResponse>()).Item;
}
```

5. Removed all Synchronous methods. Only Async now.
6. Added a logging implementation in
7. Removed all inline mocking tests and use a in memory mock server allowing you to simulate the end to end tests
8. Removed Request and Response from the API... these are internal concerns.
9. Bulked out the API alot, you should see a lot more methods on the APIs. Intern, I have brought a couple API's together and plurialized them based on the Zendesk documentation.
10. Removed the acceptance test project. It wasent really representative of acceptance tests, 'persay' and they were all covered in the new integration tests kungfu.

## DI

To register this in your DI, you just need to call...
```c#
services.AddZendeskClient("enpointurl", "username", "token");
```
then you can inject in IZendeskClient anywhere you want to use it.

if you want to add a timeout, you can do....
```c#
services.Configure<ZendeskOptions>(options => {
    options.Timeout = System.TimeSpan.FromDays(1);
});
```

NOTE: Updated description.

### CODE REVIEW

**Comment by Jetski5822** [comment]
Okay so to register this in your DI, you just need to call...
```
services.AddZendeskClient("enpointurl", "username", "token");
```
then you can inject in IZendeskClient everywhere.

File: src/ZendeskApi.Client/Resources/UploadResource.cs [hunk 1]
```diff
@@ -1,40 +1,44 @@
...
 
-        public IResponse<Upload> Post(UploadRequest request)
+        public async Task<IResponse<Upload>> PostAsync(UploadRequest request)
         {
-            var requestUri = Client.BuildUri(ResourceUri, $"filename={request.Item.FileName}{request.Token ?? ""}");
-            return Client.PostFile<UploadResponse>(requestUri, request.Item);
+            using (var client = _apiClient.CreateClient("/"))
+            {
+                var response = await client.PostAsJsonAsync($"{ResourceUri}?filename={request.Item.FileName}{request.Token ?? string.Empty}", request).ConfigureAwait(false);
+                return await response.Content.ReadAsAsync<UploadResponse>();
```
Comment (kbrimble) [inline src/ZendeskApi.Client/Resources/UploadResource.cs]: `response.Content.ReadAsAsync` won't actually upload the file here, it's just getting a response for posting a file name.

File: src/ZendeskApi.Client.Tests/Resources/AssignableGroupResourceFixture.cs [hunk 2]
```diff
@@ -1,49 +0,0 @@
-﻿using System;
```
Comment (Jetski5822) [inline src/ZendeskApi.Client.Tests/Resources/AssignableGroupResourceFixture.cs]: Im totally working on that... I didnt like the previous tests, as they wernt really validating the responses.... What they were doing was to validate that a long turned to an int and that got passed to a method.

Take a look at the new tests. The approach is to replace the old tests with true in memory end to end tests.

File: src/ZendeskApi.Client.Tests/RequestDelegateRouteBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (Jetski5822) [inline src/ZendeskApi.Client.Tests/RequestDelegateRouteBuilderExtensions.cs]: Its included in preview packages. when it comes out of preview we can include it :)

File: src/ZendeskApi.Client.Tests/ZendeskSampleSite.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+using Newtonsoft.Json;
+using ZendeskApi.Contracts.Models;
+using ZendeskApi.Contracts.Requests;
+using ZendeskApi.Contracts.Responses;
+
+namespace ZendeskApi.Client.Tests
+{
+    public class ZendeskSampleSite : IDisposable
+    {
+        public static Action<IRouteBuilder> MatchesRequest
```
Comment (Jetski5822) [inline src/ZendeskApi.Client.Tests/ZendeskSampleSite.cs]: totally... thats the idea. I an thinking to create one per resource.

File: src/ZendeskApi.Client/Logging/ILogAdapter.cs [hunk 5]
```diff
@@ -1,15 +0,0 @@
-namespace ZendeskApi.Client.Logging
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Logging/ILogAdapter.cs]: I need to include logging scopes on the api calls. totally on the map. I would also like to include the diagnostics api too.

File: src/ZendeskApi.Client/IZendeskClientFactory.cs [hunk 6]
```diff
@@ -1,12 +0,0 @@
-﻿using System;
-using ZendeskApi.Client.Http;
-using ZendeskApi.Client.Logging;
-using ZendeskApi.Client.Serialization;
-
-namespace ZendeskApi.Client
-{
-    public interface IZendeskClientFactory
-    {
-        IZendeskClient Create(Uri baseUri, ZendeskDefaultConfiguration configuration, ISerializer serializer = null, IHttpChannel httpChannel = null, ILogAdapter logger = null);
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/IZendeskClientFactory.cs]: The only thing I see of use on this API is 'maybe' ISerializer... which maybe we could introduce back in to IOptions interface.

The rest I feel is covered by the Options abstraction

File: src/ZendeskApi.Client/Resources/HttpRequestExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
+﻿using System;
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Resources/HttpRequestExtensions.cs]: The HttpChannel class I felt was a bit of a mess. now if you want to customise a HttpClient, you can just implement IZendekApiClient and return a new HttpClient. I dont see why someone would want to customise that. Is there a usecase you can think of or have come across?

File: src/ZendeskApi.Client/Serialization/ISerializer.cs [hunk 8]
```diff
@@ -1,9 +0,0 @@
-namespace ZendeskApi.Client.Serialization
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Serialization/ISerializer.cs]: Hmmmmm, maybe we could reintroduce this. Let me have a think about this one.

File: src/ZendeskApi.Client/Configuration/Headers.cs [hunk 9]
```diff
@@ -1,45 +0,0 @@
-﻿using System.Collections;
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Configuration/Headers.cs]: Hmmm... this can be done in IZendeskApiClient when building a Client, no?

**Comment by Jetski5822** [comment]
@kbrimble  Thanks for taking a look. Sorry there's alot of code movement.

Okay... so next steps are....

1. Reintroduce new tests to replace the old tests, piece mill using the new End2End site stuff.

2. Introduce Logging. I see us using 

```
using(_logger.StartScope("<method name>") {
 Do Call
}
```

3. Sketch of a ISerializer implementation... however... if this is the case, then we should have two implementations bundled,

the tests are what I am looking at, at the moment. As you can see from my last couple commits.

File: src/ZendeskApi.Client/Configuration/Headers.cs (hunk 9 above)
Comment (kbrimble) [inline src/ZendeskApi.Client/Configuration/Headers.cs]: `ZendeskClientApi` only has one argument for `CreateClient` which is for the resource. I know that you can change the headers on the resulting `HttpClient` but you never see those clients as a user as they are wrapped by a given resource.

File: src/ZendeskApi.Client/Resources/HttpRequestExtensions.cs (hunk 7 above)
Comment (kbrimble) [inline src/ZendeskApi.Client/Resources/HttpRequestExtensions.cs]: I have to say that I do agree with `HttpChannel` being a bit of a mess. I didn't originally work on this client and I can't quite see the use case either. Perhaps we just sweep this under the rug for now 😉

**Comment by kbrimble** [comment]
@Jetski5822 the steps that you mentioned make sense. For 3. I think one default implementation will be enough (probably JSON.Net). I'm interested in seeing how the logging implementation will look in amongst the actual code.

Have you found a way to do the uploads? That's basically the only bit of functionality that we've not covered in previous comments.

Let me know when you want me to take another look when you've made the rest of the changes that you want to 😺

**Comment by Jetski5822** [comment]
@kbrimble Could you ask if this works...

https://github.com/justeat/ZendeskApiClient/blob/86f65771269f572d7825912483b881c19c1a715e/src/ZendeskApi.Client/Resources/UploadResource.cs#L26

From what I can see... that endpoint doesnt exist?? https://developer.zendesk.com/rest_api/docs/core/attachments

**Comment by Jetski5822** [comment]
Hey @kbrimble - would you mind taking another look please. Think I am about done, have one or two tests left, but I am done in terms of API changes.

I removed publicly exposing the request and response objects, it made no sense felt like the internal api was bleeding out.

I have not added the ISerializer back in... as the MVC team are using Newtonsoft Json directly, I see no point in this API supporting more than one.

Nick

File: src/ZendeskApi.Build/ZendeskApi.Commons.props [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <PropertyGroup>
```
Comment (Stormsys) [inline src/ZendeskApi.Build/ZendeskApi.Commons.props]: We should re-add the metadata
```xml
    <PackageLicenseUrl>http://www.apache.org/licenses/LICENSE-2.0</PackageLicenseUrl>
    <RepositoryUrl>https://github.com/justeat/ZendeskApiClient</RepositoryUrl>
    <PackageProjectUrl>https://github.com/justeat/ZendeskApiClient</PackageProjectUrl>
```

File: src/ZendeskApi.Client/ZendeskClient.cs [hunk 11]
```diff
@@ -1,53 +1,57 @@
...
-            Uri baseUri, 
-            ZendeskDefaultConfiguration configuration, 
-            ISerializer serializer = null, 
-            IHttpChannel httpChannel = null, 
-            ILogAdapter logger = null)
-            :base(baseUri, configuration, serializer, httpChannel, logger)
+        private readonly IZendeskApiClient _apiClient;
+        private readonly ILogger _logger;
+
+        public ZendeskClient(IZendeskApiClient apiClient, ILogger logger)
```
Comment (Stormsys) [inline src/ZendeskApi.Client/ZendeskClient.cs]: Im wondering how i feel about a required(non-optional) logger.

On one hand its best practice but on the other enforces more complex setup of the client for lighter/exploratory usage of the client. I would like to be able to just get up and running without a logger if that was my choice as a consumer - what do you think?

File: src/ZendeskApi.Build/ZendeskApi.Commons.props (hunk 10 above)
Comment (Stormsys) [inline src/ZendeskApi.Build/ZendeskApi.Commons.props]: We should also add a description to both projects - its currently auto-populated.

File: src/ZendeskApi.Contracts/Models/Ticket.cs [hunk 12]
```diff
@@ -1,114 +1,110 @@
...
+        [JsonProperty("priority")]
         public string Priority { get; set; }
 
-        [DataMember(Name = "recipient", EmitDefaultValue = false)]
+        [JsonProperty("recipient")]
         public object Recipient { get; set; }
 
-        [DataMember(Name = "group_id", EmitDefaultValue = false)]
+        [JsonProperty("group_id")]
         public int? GroupId { get; set; }
```
Comment (Stormsys) [inline src/ZendeskApi.Contracts/Models/Ticket.cs]: as per #73 we should get this changed to long, i can do this in a seperate PR though.

File: src/ZendeskApi.Client/ZendeskClient.cs (hunk 11 above)
Comment (Jetski5822) [inline src/ZendeskApi.Client/ZendeskClient.cs]: Done!

File: src/ZendeskApi.Contracts/Models/Ticket.cs (hunk 12 above)
Comment (Jetski5822) [inline src/ZendeskApi.Contracts/Models/Ticket.cs]: Saweeet... done

File: src/ZendeskApi.Build/ZendeskApi.Commons.props (hunk 10 above)
Comment (Jetski5822) [inline src/ZendeskApi.Build/ZendeskApi.Commons.props]: Nice, done

File: src/ZendeskApi.Client/ZendeskApiClient.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                resource = resource + "/";
+            }
+
+            var client = new HttpClient(handler)
+            {
+                BaseAddress = new Uri($"{_options.EndpointUri}/{resource}"),
+            };
+
+            // TODO: (ngm) add auth key?
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/ZendeskApiClient.cs]: Already done - removed the comment

File: README.md [hunk 14]
```diff
@@ -1,35 +1,26 @@
...
-```
-## Accessing resources:
-```csharp
-client.Tickets...
-client.Organizations...
-client.Users...
-client.Groups...
+To register this in your DI, you just need to call...
+```c#
+services.AddZendeskClient("enpointurl", "username", "token");
```
Comment (Jetski5822) [inline README.md]: Nice, copied your example.

File: src/ZendeskApi.Client/Resources/SearchResource.cs [hunk 15]
```diff
@@ -1,34 +1,36 @@
...
-        public async Task<IListResponse<T>> FindAsync<T>(IZendeskQuery<T> zendeskQuery) where T : IZendeskEntity
+        public async Task<IListResponse<T>> Search<T>(IZendeskQuery<T> zendeskQuery)
         {
-            var requestUri = _client.BuildUri(SearchUri, zendeskQuery.BuildQuery());
-
-            return await _client.GetAsync<ListResponse<T>>(requestUri).ConfigureAwait(false);
+            using (_loggerScope(_logger, "Search"))
+            using (var client = _apiClient.CreateClient(SearchUri))
+            {
+                var response = await client.GetAsync($"{SearchUri}?{zendeskQuery.BuildQuery()}").ConfigureAwait(false);
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Resources/SearchResource.cs]: fixed, will get a test added

File: src/ZendeskApi.Client/Resources/SearchResource.cs [hunk 16]
```diff
@@ -1,34 +1,36 @@
...
         {
-            var requestUri = _client.BuildUri(SearchUri, zendeskQuery.BuildQuery());
-
-            return _client.Get<ListResponse<T>>(requestUri);
+            _apiClient = apiClient;
+            _logger = logger;
         }
 
-        public async Task<IListResponse<T>> FindAsync<T>(IZendeskQuery<T> zendeskQuery) where T : IZendeskEntity
+        public async Task<IListResponse<T>> Search<T>(IZendeskQuery<T> zendeskQuery)
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Resources/SearchResource.cs]: done

**Comment by Jetski5822** [comment]
@Stormsys Thanks for the code review, ive addressed most comments.

In terms of the Search API, I havent really looked at it. I pushed some fixes up, but looking at the API docs, https://developer.zendesk.com/rest_api/docs/core/search - It looks like we could do better.

https://support.zendesk.com/hc/en-us/articles/203663226

What do you think?

**Comment by Jetski5822** [comment]
One thing to note is that I removed Pagination for list apis.... this was a mistake as I couldnt see it in their API. So need to add this back.

https://developer.zendesk.com/rest_api/docs/core/introduction#pagination

**Comment by Stormsys** [comment]
@Jetski5822 Regarding the search api how do you see it improving? we currently use it like this as an example:
```c#
client.Search.Find(new ZendeskQuery<User>().WithCustomFilter("email", email))
```

`WithCustomFilter` also supports the operators input.

**Comment by Jetski5822** [comment]
@Stormsys not sure.... thinking something along the lines of...

_client.SearchAsync<User>(query => query.WithFilter("email", user.email));

I think a fluent interface for search makes things much cleaner. Im working on Pagination at the moment. Its looking really nice.

**Comment by Jetski5822** [comment]
@Stormsys Checkout the changes I made to the Search API and syntax changes, what do you think?

**Comment by Stormsys** [comment]
@Jetski5822 i like it. but i think we lost .WithPaging and CustomFields

**Comment by Jetski5822** [comment]
@Stormsys  For Paging, you now can just use the paging parameters which is used the same on all list endpoints (Paging is optional). CustomFields now is called Fields

I meant to remove the SearchAsync generic type, its restricting, and we want this api as open as possible based on the api spec.

Nooo the Action totally needs to be required, we should probably throw an error if no filters are applied.

Accept Header issue fixed, thanks for spotting.

Need to think about the other things......

**Comment by Jetski5822** [comment]
@Stormsys can you give me some examples of what you are seeing returned via the search API? Just change some of the data to hide any PPI data.

According to the spec https://developer.zendesk.com/rest_api/docs/core/search No additional items are returned!?

**Comment by Stormsys** [comment]
@Jetski5822 good work! 

I missed the pagination input, okay it makes sense im happy with that.

Should we remove BuildQuery from the interface? `.SearchAsync(q => q.BuildQuery())` feels like something you should't be able to do, but maybe it doesn't matter also ^^.

As for sample response theres quite a lot for me to obsficate but the basic gist is this, imgine you took your `SearchResult` and merged it with the type that corrisponds to `ResultType` so where ResultType = ticket, the object would be `SearchResultTicket : SearchResult, Ticket` and where ResultType=organisation, the object would be `SearchResultOrganization : SearchResult, Organization`  - i hope that made sense. so the results object can be a ticket, orginisation or any other resource that has the extra fields specified in the search api(listed in `SearchResult`)

If you still need a sample let me know and il see what i can do.

**Comment by Jetski5822** [comment]
Ohhhhhh I think I see what you are talking about, yes, "results" can be polymorphic... however, its representation is not just one type. The documentation is not very clear on this and I only just kinda understand it.

So.... To fix this... I have an idea.... but what it will mean is that the SearchAsync Api will change to

> async Task<IPagination<object>> SearchAsync(Action<IZendeskQuery> builder, PagerParameters pager = null)

To then use this you will need to do... 

> await _client.SearchAsync(etc etc).TypeOf<Ticket>();

This is the most generic I think I can make it... not sure....

**Comment by Jetski5822** [comment]
Oh the BuildQuery thing.... yeah, Ive been wrestling with whether to remove or not... still not sure.

**Comment by Stormsys** [comment]
Wont we lose the identifiying information (ResultType string)?

I think if we do this we should provide helper methods to convert the type, or atleast  extensions methods that can give .Where(result => result.ResultType == "ticket").Select(result => result.TypeOf(Tickets))

**Comment by Jetski5822** [comment]
Yes, that would suck.

Let me sleep on it, I'll get back to you tomorrow. N

**Comment by Stormsys** [comment]
By the way you said: CustomFields now is called Fields in my response object for tickets "fields" and "custom_fields" are two distinct entries.

Alright, Good Job though and thanks for the time you have been putting into this!

**Comment by Jetski5822** [comment]
Sorry I meant, I replaced the WithCustomFilter with the name WithFilter as the underlying class is called Filter.

I dont think we will miss anything if we lose the ResultType, is there a reason you need it?

**Comment by Stormsys** [comment]
No as long as we have the original type it's enough- from the way I've seen it used anyway.

**Comment by Stormsys** [comment]
I think we should definitely provide on the response class helpers for the possible return types(tickets, organisations, users and groups) - it might be a little unobvious api wise if we don't provide these to convert the types back out.

**Comment by Jetski5822** [comment]
@Stormsys Just pushed an update for Search, Should provide all the stuff you need. what do you think?

**Comment by Stormsys** [comment]
@Jetski5822 theres currently a bug in `ZendeskApi.Client.Converters.SearchJsonConverter.ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)` : `InvalidCastException: Invalid cast from 'System.String' to 'System.Uri'.`

**Comment by Jetski5822** [comment]
Could you past the string that's causing the issue?

**Comment by Jetski5822** [comment]
@Stormsys Fixed that bug, seems Value<Uri> doesnt work for some reason.

**Comment by Stormsys** [comment]
Looks like its working.

What do you think about adding a OfType to IZendeskQuery so instead of:
`.SearchAsync(q => q.WithFilter("type", "ticket", FilterOperator.Exact));`
You would have the option to do:
`.SearchAsync(q => q.OfType<Ticket>().WithFilter(...));` just for convinience

**Comment by Jetski5822** [comment]
Oh, dont do that... you can do this.....

```
.SearchAsync(q => q.WithTypeFilter<Ticket>());
```
or event better
```
.SearchAsync<Ticket>(q => {});
```

**Comment by Stormsys** [comment]
Yup either of those work for me.

How ready and stable do you feel the PR is for merging? i will have a chat with @kbrimble about merging for a -pre as well as testing it out in the field myself.

**Comment by Jetski5822** [comment]
@Stormsys I think its ready, and anything else can be a PR afterwards.

Though the title of the PR is an upgrade to netstandard1.4, its basically a complete rewrite., not sure if any of the original code survived! :)

**Comment by Jetski5822** [comment]
@Stormsys Saweeet. Let me know how it goes.

**Comment by Jetski5822** [comment]
@Stormsys I have 1 more thing Id like to push if thats okay?

**Comment by Jetski5822** [comment]
Actually decided it was worth it.... Enjoy netstandard1.3 :)

**Comment by slang25** [comment]
@Jetski5822 This is fantastic, thanks for your contribution!

**Comment by Jetski5822** [comment]
@slang25 yw 👍

**Pull request merged**